pkg/chunk/cache/redis_client.go

```diff
@@ -0,0 +1,309 @@
+	f.StringVar(&cfg.Endpoint, prefix+"redis.endpoint", "", description+"Redis service endpoint to use when caching chunks. If empty, no redis will be used.")
+	f.DurationVar(&cfg.Timeout, prefix+"redis.timeout", 100*time.Millisecond, description+"Maximum time to wait before giving up on redis requests.")
+	f.DurationVar(&cfg.Expiration, prefix+"redis.expiration", 0, description+"How long keys stay in the redis.")
+	f.IntVar(&cfg.PoolSize, prefix+"redis.pool_size", 0, description+"Maximum number of socket connections in pool.")
+	f.Var(&cfg.Password, prefix+"redis.password", description+"Password to use when connecting to redis.")
+	f.BoolVar(&cfg.EnableTLS, prefix+"redis.enable-tls", false, description+"Enables connecting to redis with TLS.")
+	f.DurationVar(&cfg.IdleTimeout, prefix+"redis.idle-timeout", 0, description+"Close connections after remaining idle for this duration. If the value is zero, then idle connections are not closed.")
+	f.DurationVar(&cfg.MaxConnAge, prefix+"redis.max_conn_age", 0, description+"Close connections older than this duration. If the value is zero, then the pool does not close connections based on age.")
+
+	f.IntVar(&cfg.DeprecatedMaxIdleConns, prefix+"redis.max-idle-conns", 0, "Deprecated: "+description+"Maximum number of idle connections in pool.")
+	f.IntVar(&cfg.DeprecatedMaxActiveConns, prefix+"redis.max-active-conns", 0, "Deprecated (use pool_size instead): "+description+"Maximum number of active connections in pool.")
+	f.DurationVar(&cfg.DeprecatedMaxConnLifetime, prefix+"redis.max-conn-lifetime", 0, "Deprecated (use max_conn_age instead): "+description+"Close connections older than this duration.")
+	f.BoolVar(&cfg.DeprecatedWaitOnPoolExhaustion, prefix+"redis.wait-on-pool-exhaustion", false, "Deprecated: "+description+"Enables waiting if there are no idle connections. If the value is false and the pool is at the max_active_conns limit, the pool will return a connection with ErrPoolExhausted error and not wait for idle connections.")
+}
+
+// Validate Redis configuration
+func (cfg *RedisConfig) Validate() error {
+	switch cfg.Topology {
+	case redisTopologyServer, redisTopologyCluster, redisTopologySentinel:
+	default:
+		return fmt.Errorf("unsupported Redis topology %q", cfg.Topology)
+	}
+	return nil
+}
+
+// RedisClient is a generic Redis client interface
+type RedisClient interface {
+	Ping(context.Context) error
+	MSet(context.Context, []string, [][]byte) error
+	MGet(context.Context, []string) ([][]byte, error)
+	Close() error
+}
+
+type redisCommander interface {
+	Ping(ctx context.Context) *redis.StatusCmd
+	TxPipeline() redis.Pipeliner
+	MGet(ctx context.Context, keys ...string) *redis.SliceCmd
+	Close() error
+}
+
+var (
+	_ RedisClient = (*redisBasicClient)(nil)
+	_ RedisClient = (*redisSentinelClient)(nil)
+
```

Comment:
These two lines are not necessary, compiler will report problems in NewRedisClient function.